### PR TITLE
Fix concurrency issues between cancelling a relocation and marking shard as relocated

### PR DESCRIPTION
**Pull request by bleskes**
Once a primary is marked as relocated, we can not safely move it back to started (as we have no way of waiting on inflight operations that are performed on the target primary). If the master cancels the relocation in that state, we fail the primary. Sadly, there is a racing condition between the `updateRoutingEntry` method (which is called when the relocation is cancelled by the master) and the `relocated` method. That racing condition can leave the shard as marked "relocated" but have the routing entry not reflect the target relocation. This in turns causes NPEs in TransportReplicationAction:

```
java.util.Objects requireNonNull Objects.java 203
org.elasticsearch.action.support.replication.TransportReplicationAction$ConcreteShardRequest <init> TransportReplicationAction.java 982
```

Sadly, once we end up in this state, we will never recover.

This commit fixes that race condition by making sure `updateRoutingEntry` acquires the mutex when checking for the relocated status.  While at it, I also tightened up the code and added lots of assertions/hard checks.

### CODE REVIEW

**Comment by bleskes** [comment]
@s1monw can you take a look?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -368,60 +366,46 @@ public QueryCachingPolicy getQueryCachingPolicy() {
...
+            persistMetadata(newRouting, currentRouting);
         }
-
-        if (state == IndexShardState.RELOCATED &&
-            (newRouting.relocating() == false || newRouting.equalsIgnoringMetaData(currentRouting) == false)) {
-            // if the shard is marked as RELOCATED we have to fail when any changes in shard routing occur (e.g. due to recovery
-            // failure / cancellation). The reason is that at the moment we cannot safely move back to STARTED without risking two
-            // active primaries.
-            throw new IndexShardRelocatedException(shardId(), "Shard is marked as relocated, cannot safely move to state " + newRouting.state());
+        if (currentRouting != null && currentRouting.active() == false && newRouting.active()) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: should we also synchronize the notification on the `indexEventListener` maybe to ensure they are not called our of order?

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: maybe that's an overkill

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: nevermind this is called by the clusterstate updatae thread :)

**Comment by s1monw** [comment]
LGTM

**Comment by bleskes** [comment]
thx @s1monw

**Pull request merged**